### PR TITLE
fix truncated lines issue

### PR DESCRIPTION
**Pull request by jesseduffield**

### CODE REVIEW

**Review by ed-bassett — CHANGES_REQUESTED** [review]

File: src/parse/mod.rs [hunk 1]
```diff
@@ -28,27 +28,40 @@ pub fn parse(text: String) -> Vec<Row> {
   text
     .lines()
     .map(|line| {
-      let mut cells = vec![];
-
-      let mut last = 0;
       // I want to get the chars as an array, then slice that up.
       let chars = line.chars().collect::<Vec<char>>();
```
[the remark this answers is not in the document]
Comment (jesseduffield) [reply src/parse/mod.rs:32]: Ah, very nice

File: src/parse/mod.rs [hunk 2]
```diff
@@ -28,29 +33,29 @@ pub fn parse(text: String) -> Vec<Row> {
...
 
-      for position in &column_indices[1..] {
-        push(&mut cells, chars[last..*position].to_owned());
-        last = *position;
-      }
-      push(&mut cells, chars[last..].to_owned());
+      let char_len = chars.len();
```
Comment (ed-bassett) [inline src/parse/mod.rs:39]: Actually, on thinking this through a bit more, if we know the sizes of the columns ahead of time, we can just do this with iterators:
```
  let column_sizes = get_column_indices(&text).into_iter().tuple_windows().map(|(i0, i1)| i1-i0).chain(once(usize::MAX)).collect::<Vec<_>>();

  text
    .lines()
    .map(|line| {
      // I want to get the chars as an array, then slice that up.
      let cells = column_sizes.iter().scan(line.chars(), |chars, column_size|
        Some(chars.take(*column_size).collect::<String>().trim_end().to_owned())
      ).collect();

      Row::new(line.to_owned(), cells)
    })
    .collect()
```

Comment (jesseduffield) [reply src/parse/mod.rs:39]: I'm not quite sure what's happening there: is the `chars` from the scan closure being mutated? If not, how are we managing to get the right chars when we use `.take`? If my column sizes are [3, 3, 3], I would expect this code to return the same three characters on each iteration of the loop

**Review by ed-bassett — APPROVED** [review]

File: src/parse/mod.rs (hunk 2 above)
Comment (ed-bassett) [reply src/parse/mod.rs]: `scan` is essentially a `map` with some well-defined statefulness in the form of a mutable reference to whatever's passed as the first argument. In this case, we're passing `line.chars()` which is itself an iterator, so as we iterate through `column_sizes` we can mutate the `line.chars()` iterator. I.e. for each `column_size` we `take` that many characters from `line.chars()`, consuming part of that iterator.

Comment (jesseduffield) [reply src/parse/mod.rs]: That's very interesting: it's halfway between a map and a reduce. Begs the question of whether there is an analogous function for 'filter' with statefulness as well

**Pull request merged**